README.md

```diff
@@ -1,2 +1,79 @@
 # vm-device
-A virtual machine device model crate
+
+The `vm-device` crate provides:
+* device traits defining read and write operations on specialized buses
+* device manager (bus-specific traits and a concrete implementation) for operating devices and dispatching I/O
+* abstractions for defining resources and their constraints (e.g. a specific bus address range, IRQ number, etc)
+
+## Design
+
+The virtual device model is built around four traits, `DevicePio` and `MutDevicePio` for [Programmed I/O](https://en.wikipedia.org/wiki/Programmed_input%E2%80%93output) (PIO), and `DeviceMmio` and `MutDeviceMmio` for [Memory-mapped I/O](https://en.wikipedia.org/wiki/Memory-mapped_I/O) (MMIO). The traits define the same methods for handling read and write operations. The difference is that `DevicePio` and `DeviceMmio` only require immutable self borrows, whereas `MutDevicePio` and `MutDeviceMmio` require mutable borrows.
+
+The device manager abstraction is implemented by the `IoManager` struct. It defines two buses, one for PIO and one for MMIO. For each bus, with the help of the `PioManager` and `MmioManager` traits, the manager provides methods for device registration, as well as for dispatching read and write requests. The manager will determine which device is responsible for handling the I/O request based on the accessed address range, and will route the request to that device.
+
+`PioManager` and `MmioManager` traits are useful as interfaces for a couple of reasons. First, to allow for alternative implementations when the provided `IoManager` is not sufficient. Second, to allow certain parts of a VMM to make use of and depend on a particular type of I/O without bringing the whole `IoManager` is scope. 
```

Comment:
Instead of the "VMM to make use of ...." we can also say that it allows for other crates to depend on the Managers without depending on an actual implementation. The vm-device is going to be a dependency in the context of vm-pci, so in that case even other rust-vmm crates might have dependencies on vm-device.